cpp/include/cudf/detail/gather.cuh

```diff
@@ -387,16 +378,22 @@ gather(table_view const& source_table, MapIterator gather_map_begin,
 
   for(auto const& source_column : source_table) {
     // The data gather for n columns will be put on the first n streams
-    destination_columns.push_back(
-                                  cudf::experimental::type_dispatcher(source_column.type(),
-                                                                      column_gatherer{},
-                                                                      source_column,
-                                                                      gather_map_begin,
-                                                                      gather_map_end,
-                                                                      nullify_out_of_bounds,
-                                                                      mr,
-                                                                      stream));
+    auto dest = cudf::experimental::type_dispatcher(source_column.type(),
+                                                    column_gatherer{},
+                                                    source_column,
+                                                    gather_map_begin,
+                                                    gather_map_end,
+                                                    nullify_out_of_bounds,
+                                                    mr,
+                                                    stream);
+
+    // Allocate null mask if we need one
+    if (source_column.nullable() or nullify_out_of_bounds) {
+      auto mask = create_null_mask(dest->size(), mask_state::ALL_NULL, stream, mr);
```

Comment:
Nit: Is it necessary to clear the mask here?  If I'm reading the code correctly, it's going to unconditionally gather the source masks into these mask buffers immediately afterwards.  I think we can avoid a memset invocation here.

Reply:
There's a parameter `nullify_out_of_bounds` that allows the gather map to contain invalid indices, in which case those rows are skipped and left as nulls. I suppose I could modify the `gather_bitmask` kernel to allow explicitly setting those rows to null instead of relying on memset.

Reply:
Thanks for looking into it.  Not a must-fix at all, I just thought it could be avoided.  If it's complicated or risky to optimize, may be best handled as a followup issue.

Reply:
Fixed